runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
> So externalFds is all the file descriptors, but ExtraFiles is everything 2 and up.

yes

> This isn't a libcontainer spec ;).

I am not an author of this pull request;))

> I can imagine an OCS runtime that
supports sparse FD arrays, and it's trivial to support in the spec, so
why not allow it here?  If it's too much work for libcontainer to
migrate their array to a map (which seems unlikely), just have
libcontainer explain that it adds an additional continuity requirement
to the externalFds keys.

I think your understanding of externalFds is incomplete. The index in this array isn't interested for us. Only the order of elements is impotant. We generate two arrays of externalFds. The first one is generated when a container is started, and the second one is generated when a container is restored. We use these array to map elements from the first array to the second array (f[i] -> s[i]).
ExternalFds can be continuous even when FD array is sparce.

> Who's we?  The runtime?  CRIU?  It's probably not going to be the application ;).

CRIU will restore a proper end of pipe for each descriptor.